app.py

```diff
@@ -0,0 +1,162 @@
+import os
+
+from pyramid.config import Configurator
+from pyramid.view import view_config
+from pyramid.httpexceptions import HTTPFound
+
+from pyramid.security import remember, forget
+from pyramid.authentication import AuthTktAuthenticationPolicy
+from pyramid.authorization import ACLAuthorizationPolicy
+from cryptacular.bcrypt import BCRYPTPasswordManager
+
+import sqlalchemy as sa
+from sqlalchemy.orm import scoped_session, sessionmaker
+from zope.sqlalchemy import ZopeTransactionExtension
+from sqlalchemy.ext.declarative import declarative_base
+
+from waitress import serve
+
+HERE = os.path.dirname(os.path.abspath(__file__))
+DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension()))
+DATABASE_URL = os.environ.get(
+    'DATABASE_URL',
+    'postgresql://power_user:hownowbrownsnake@localhost:5432/test1'
+    #'postgresql://power_user:nopassword@localhost:5432/test1'
+)
+Base = declarative_base()
+
+
+class User(Base):
+    __tablename__ = 'users'
+
+    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
+    username = sa.Column(sa.Unicode(127), nullable=False, unique=True)
+    password = sa.Column(sa.Unicode(127), nullable=False)
+
+    @classmethod
+    def new(cls, username=None, password=None, session=DBSession):
+        manager = BCRYPTPasswordManager()
+        if not (username and password):
+            raise ValueError("Username and password needed")
+        hashed = manager.encode(password)
+        instance = cls(username=username, password=hashed)
+        session.add(instance)
+        return instance
+
+    @classmethod
+    def get_by_id(cls, id, session=DBSession):
+        return session.query(cls).filter(cls.id == id).one()
+
+    @classmethod
+    def get_by_username(cls, username, session=DBSession):
+        return session.query(cls).filter(cls.username == username).one()
+
+
+@view_config(route_name="login", renderer="templates/gatepage.jinja2")
+def login_page(request):
+    if request.method == "POST":
+        username = request.params.get('username', '')
+        authenticated = False
+        try:
+            authenticated = login(request)
```

Comment:
This function returns `None` or raises an error.  What would happen if it returned `username` instead, so that you didn't have to pull that value out of the `request` twice.